### PR TITLE
#2 Template datasource

### PR DESCRIPTION
**Pull request by ja6a**
This allows an environment to reference by its name template's name. Previously the ID had to be known.

```
data "skytap_template" "foo" {
	name = "CentOS 6.10 Desktop Firstboot"
}

output "id" {
  value = "${data.skytap_template.foo.id}"
}

resource "skytap_environment" "bar" {
	template_id = "${data.skytap_template.foo.id}"
	name = "my environment"
	description = "This is my environment"
}
```

### CODE REVIEW

**Comment by jchenry** [comment]
I think we're going to have issues here. Names are not a globally unique Identifier in the system. Templates, for better or worse, can have duplicate names.

**Comment by pegerto** [comment]
Hello @jchery.

We agree in this statement, the objective of the PR is to provide a placeholder to enable the user options to filter the template in a descriptive manner. And then use the id as identification. The current use case only allows name and it is mandatory. This interface is not sufficient. 

We should discuss the possible filters, including name tags, etc.

Now for the scope of this PR we should define the minimum, at this stage I think it should return a list and the name to be optional, or even a regex expression of the name.

File: skytap/datasource_skytap_template.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+		return fmt.Errorf("error retrieving templates: %s", err)
+	}
+
+	templates := filterDataSourceSkytapTemplatesByName(templatesResult.Value, name)
+
+	if len(templates) == 0 {
+		return fmt.Errorf("no template found with name %s", name)
+	}
+
+	if len(templates) > 1 {
```
Comment (pegerto) [inline skytap/datasource_skytap_template.go]: I believe this is valid, not sure if getting the list of then in the error will be clear,  review the use case in AWS where, 

https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/data_source_aws_ami.go#L260-L272

There is an attribute most_recent

File: skytap/datasource_skytap_template.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+func dataSourceSkytapTemplate() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceSkytapTemplateRead,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				Description:  "name of the template",
```
Comment (pegerto) [inline skytap/datasource_skytap_template.go]: use a regex expression instead, in combination with most_recent
```
	"most_recent": {
	Type:     schema.TypeBool,
       Optional: true,
       Default:  false,
       ForceNew: true,
},
```

**Comment by pegerto** [comment]
Hi @jchenry

A standard behaviour of resolving conflicts will be the use of the latest, that is a flag, and will be added to this pull request. The use of a full name isn't flexible therefore we are using a regex attribute for selecting the template by name.

Additional criteria as tags, owner etc, will be add in latter tickets if the time allows it.

File: skytap/resource_skytap_environment_test.go [hunk 3]
```diff
@@ -134,8 +134,12 @@ func testAccCheckSkytapEnvironmentDestroy(s *terraform.State) error {
 
 func testAccSkytapEnvironmentConfig_basic(rInt int) string {
 	return fmt.Sprintf(`
+data "skytap_template" "foo" {
```
Comment (pegerto) [inline skytap/resource_skytap_environment_test.go]: Move this acceptance test, not from the skytap environment, but into a specific data source acceptant test.

**Pull request merged**